Add module to gather facts of a SolidFire cluster.

hulquest opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Add module to gather facts of a SolidFire cluster.  The module gathers:
- Cluster information
- List cluster nodes.
- Cluster state
- Accounts
- Drives
- Volumes
- Virtual Volumes (vvols)
- Snapshots
- Cluster Faults
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
ansible/modules/storage/netapp/sf_facts.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
✔ ~/src/projects/ansible [sf-facts|✚ 1…2] 
14:44 $ ansible --version
ansible 2.4.0 (sf-facts de9f032a10) last updated 2017/09/01 14:55:21 (GMT -600)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/khulques/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/khulques/src/projects/ansible/lib/ansible
  executable location = /home/khulques/src/projects/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Akasurde commented on lib/ansible/modules/storage/netapp/sf_facts.py:
```diff
@@ -0,0 +1,149 @@
+
+# (c) 2017, NetApp, Inc
+#
+# This file is part of Ansible
```
Can you please replace line 4 to 19 to match a single liner GPL license (which is new in Ansible Module development)? 


```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

Akasurde commented on lib/ansible/modules/storage/netapp/sf_facts.py:
```diff
@@ -0,0 +1,149 @@
+
+import ansible.module_utils.netapp as netapp_utils
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
```
get_exception is deprecated in favor of Python native exception handling. 

```
except Exception as e:
    module.fail_json(msg=to_native(e)
```

Akasurde reviewed (changes requested):
Rest Looks good to me.

lmprice commented on lib/ansible/modules/storage/netapp/sf_facts.py:
```diff
@@ -0,0 +1,138 @@
+        """Gather information from the cluster and return it to Ansible."""
+
+        self.facts['mvip'] = self.module.params['hostname']
+        """This is the cluster identifyer."""
```
Typo here.

lmprice reviewed (approved):
Very minor nit, otherwise, LGTM.
shipit

Akasurde commented on lib/ansible/modules/storage/netapp/sf_facts.py:
```diff
@@ -0,0 +1,138 @@
+short_description: Report facts of a NetApp SolidFire cluster.
+extends_documentation_fragment:
+    - netapp.solidfire
+version_added: '2.4'
```
2.4 branch is freezed for development so I guess version_added will be 2.5

Take a look at - [shippable](https://app.shippable.com/github/ansible/ansible/runs/36453/1/console)

broncofan commented:
LGTM, 
shipit

Akasurde reviewed (approved):
shipit

resmo commented on lib/ansible/modules/storage/netapp/sf_facts.py:
```diff
@@ -0,0 +1,138 @@
+"""
+
+RETURN = """
+msg:
```
msg is a common return and must not be documented. But your returns should be identical to the dict of facts returned.

bcoca committed:
move from with_<lookup>: to loop:

- old functionality is still available direct lookup use, the following are equivalent

  with_nested: [[1,2,3], ['a','b','c']]

  loop: "{{lookup('nested', [1,2,3], ['a','b','c'])}}"

- avoid squashing with 'loop:'
- fixed test to use new intenal attributes
- removed most of 'lookup docs' as these now reside in the plugins

bcoca committed:
config tests

also a couple of fixes to manager

dj4ngo committed:
[feat] module apt: add warning on deprecated states

bcoca committed:
updated dig lookup for bot

bcoca committed:
remove unused ssh pipelining setters

fixes #31125

bcoca committed:
report extra info in verbose and if needed

Commit:
Fix backwards compatibility of constants.get_config

get_config would use ConfigManager.get_ini_value which does not
exist. What we are meant to use is
ansible.config.manager.get_ini_config_value and this method does not
expect a list, only a dictionary with a section and a key.

bcoca committed:
added toggle to control vars plugin behviour

this should allow user to control how they want the playbook dirs inspected
for additional vars, default now reverts to 2.3 behaviour (top).
corrected paths order
minor doc reword

alikins committed:
Fix flush_cache on redis cache

Fix a del on non existent self.cache.

Fixes #31148

caphrim007 committed:
Adds the bigip_remote_syslog module

This patchs add a remote-syslog module that can be used by people
to manage their (basic, not HSL) syslog configuration

dannyroberts committed:
fix monit version check

it was erroneously checking > 3.18 instead of > 5.18

Fixes #30614

bcoca committed:
better chown errors

fgbulsoni committed:
LogicalInterconnectGroupFactsModule for HPE OneView  (#28847)

* Added support to retrieving LIG resources in HPE OneView

* Fixing copyright header according to review

* Swapping out config for full credentials in parameter for documentation

wimnat committed:
[cloud] Only get rules if listener in elb_application_lb been identified in compare_rules (#30604)

* Only get rules if listener has been identified in compare_rules

* Always cast the listener port to an integer.

yeukhon committed:
Fixes #30073: Remove 'Consider hg command' warning. (#31159)

This PR addresses two issues:

1. The hg module was added to command module's check_command list,
so if someone runs hg directly from the command module, the command
module would warn the user "Consider using hg module rather than running hg".

We address this by removing hg from the list.

2. We added a new note to tell users push feature will be addressed
in issue #31156.

clasohm committed:
add https_proxy (#31062)

* add https_proxy

* fix indentation

* use http.nonProxyHosts

shuklaabhi committed:
BugFix code breaks, if include_vars dir is file instead of directory (#31157)

4383 committed:
Feature/test inventory (#30707)

* [test] move inventory test to right path
* [feat] add unit test for yaml inventory plugin

dagwieers committed:
foreman: PEP8 compliancy and doc fixes (#30910)

This PR includes:
- PEP8 compliancy fixes
- Documentation fixes

dagwieers committed:
rhn_channel: PEP8 compliancy and doc fixes (#30912)

This PR includes:
- PEP8 compliancy fixes
- Documentation fixes

rahushen committed:
Fixes #30769 - nxos_vtp_version errors out on N1 images. (#30770)

* Fixes #30769

* add the same fix for 2 more vtp modules

bcoca committed:
make pipelining actually backwards compatible

restore ssh specific setting the global, in the future these will be separate.

sivel committed:
Update uptime.py example script with changes to the API. Fixes #31229

bcoca committed:
set _hosts on access if None (#31111)

set _hosts on access if None to bpyass srlz10n issues to fix #30903

alikins committed:
Handle vault decrypt --output=- (#31066)

In cli.CLI.unfrack_path callback, special case if the
value of '--output' is '-', and avoid expanding
it to a full path.

vault cli already has special cases for '-', so it
just needs to get the original value to work.

Fixes #30550

alikins committed:
Fixes for facts distribution.py (#31110)

'distribution' facts were being set after checking
the existence of the dist file, and then being set
again with more detail after they were succesfully parsed.

But if the dist file was not succesfully parsed and
matched the required names, the loop continues
without resetting the earlier set facts. This is
how 'Mandriva' would end up being the 'distribution'
file for unrelated cases (it would find /etc/lsb-release,
set distro to 'Mandriva', then fail to parse/match and
continue the loop. If no other checks worked, 'Mandriva'
would stick).

* parse_dist_file_NA should check 'name' not distro for NA

parse_distribution_file_NA was checking the incoming
'distribution' fact to be 'NA', but the fact itself can
be specific at that point ('KDE Neon', for ex) but the
check is really if the 'name' it was passed is NA.

* for matches on OS_RELEASE_ALIAS (ie, 'Archlinux') do
not continue if the dist file content doesn't match. Previously
it had to because of the 'Mandriva' bug mentioned above.

This is a more general fix for #30693 than #30723

Fixes  #30693
Related to #30600

bcoca committed:
dont follow symlinks for inventories

now symlink dir is checked for locality of group/host_vars

fixes #31195

andreaso committed:
Remove git:// urls from docsite and from packaging (#30459)

It's almost always preferable to clone by way of https:// rather than
using the bare git:// protocol. Not only does https:// provide
stronger guarantees it also plays nicer with corporate proxies, etc.

mscherer committed:
Add support for LVM to filesystem module (#21620)

And since LVM2_member is not a very userfiendly name, add
a mechanism for more user friendly names.

alikins committed:
Fix ansible_distribution on Ubuntu 10.04 (#31108)

The /etc/os-release based distro detection doesn't
seem to work for Ubuntu 10.04 (no /etc/os-release?).

So it was testing the next case which was /etc/lsb-release to
see if it is 'Mandriva'. Since the check for existence of
(/etc/lsb-release, Mandrive) was the first non-empty dist
file match, 'ansible_distribution' was being set to 'Mandriva'
expecting to be corrected by the data from the dist file content.

But since the dist file parsing for Mandriva didn't match for
Ubuntu 10.04 /etc/lsb-release _and_ there is no Debian specific
lsb-release check, 'ansible_distribution' stayed at 'Mandriva'
and the dist file checking loop keeps going and eventually off
the end of the list before finding a better match.

Adding a debian/ubuntu specific check for /etc/lsb-release after
the debian os-release sets the info correctly and stops further
checking of dist files.

Fixes #30693

mattclay committed:
Run OS X and FreeBSD CI tests using groups. (#31254)

* Run OS X and FreeBSD CI tests using groups.
* Improve readability of Shippable jobs.

jborean93 committed:
win_copy: fix for copying encrypted file without pass (#31084)

* win_copy: fix for copying encrypted file without pass

* fix pep8 issue

* reduced the diff and fixed some minor issues

virtuald committed:
win_reg_stat actually sets properties to be a dictionary, not a list (#31252)

bcoca committed:
turn all inventory cache 'off by default'

abadger committed:
We want the ini and env vars to both override in the same order

There's an ssh version and a generic version.  We want the ssh version
to override the generic version in both cases.

mattclay committed:
Switch CI for OS X back to a single job. (#31265)

* Update Shippable scripts for group/non-group use.
* Switch CI for OS X back to a single job.

mattclay committed:
Temporarily disable OS X in CI.

jborean93 committed:
win_become: move error handling to Ansible outside of shell (#31227)

* win_become: move error handling to Ansible outside of shell

* trimmed the output so double newlines don't get set

* added test for non-zero exit code

* missed issue URL on test

* changed exit to SetShouldExit

bcoca committed:
corrected version added

samdoran committed:
Correctly write SELinux config file (#31251)

* Add new lines to end of config file lines

* Properly write out selinux config file

Change module behavior to not always report a change but warn if a reboot is needed and return reboot_required.

Improve the output messages.

Add strip parameter to get_file_lines utility to help with parsing the selinux config file.

* Add return documentation

* Add integration tests for selinux module

* Use consistent capitalization for SELinux

* Use atomic_move in selinux module

* Don't copy the config file initially

There's no need to make a copy just for reading.

* Put message after set_config_policy in case the change fails

* Add aliases to selinux tests

jvandermeulen committed:
Make YUM repo URL clickable again (#31264)

Make YUM repo URL clickable again

mattclay committed:
Revert "Temporarily disable OS X in CI."

This reverts commit b26ae1809c69497d52b8e00017360d533e3950bc.

rahushen committed:
Fixes nxos_igmp_interface idempotence issue for N1 images (#31220)

SteveHNH committed:
Add update_only parameter for yum module (#22206)

* Add update_only parameter for yum module

When using latest, `update_only: yes` will ensure that only existing
packages are updated and no additional packages are installed.

* Update yum.py

Update version added for `update_only` parameter to 2.5

* add unit tests for update_only flag in yum module

machacekondra committed:
ovirt_vms: Support to specify VM id (#31253)

Fixes: https://github.com/ansible/ansible/issues/30873

machacekondra committed:
ovirt_hosts: print error when host for iscsi login not found (#30928)

KKoukiou committed:
cloud: ovirt_datacenter.py: Add 'force' parameter for removing dc (#30352)

machacekondra committed:
ovirt_vms: Raise proper error when template isn't found (#30306)

machacekondra committed:
ovirt_clusters: Fix fencing and kuma comparision (#30302)

Commit:
Return None for DC in case of detached storage domain (#29175)

The following patch fixes a regression when trying to remove a detached
storage domain.
As part of the remove process the ovirt_storage_domains module first
tries to move the domain to maintenance and detach it.
In case of removing a detached storage domain with no DC attached to it
The maintenace process will fail with 404 (not exists) exception when
trying to fetch the DC using empty Guid.
The fix proposes a solution to return None value in case of a detached
storage domain.

KKoukiou committed:
cloud:ovirt:ovirt_storage_domains.py: Don't require extra params when state==absent (#30337)

Make sure that example in docs is usable:

 # Remove storage domain
 - ovirt_storage_domains:
     state: absent
     name: mystorage_domain
    format: true

Without this PR data_center and host parameters where required when we wanted to
remove some storage domain.

Also fixes a regression when trying to remove a detached
storage domain.

bcoca committed:
updated constructed docs to clarify use of ext

bcoca committed:
correctd inventory_file to return full path

fixes #31294

bcoca committed:
check type to avoid typeerror tb

fixes #31290 by giving more meaningful message

bcoca committed:
Revert "fix git wrapper (#29069)"

This reverts commit f8005d27379c4d67bc86cf6af83ba1b1823bb10f.

fix needs to be rethought as it applies to only newer git versions
and use of env shell breaks with non 'bourne compatible' shells

caphrim007 committed:
Adds the bigip_monitor_https module (#31205)

This patch adds the https_monitor module which allows people
to create and manage HTTPS monitors.

bcoca committed:
made note about cwd not working for hostlist anymore (#31258)

* made note about cwd not working for hostlist anymore

fixes #31041

* Edits

mattclay committed:
Fix ansible-test config management.

mkrizek committed:
yum: fix name parsing out of envra (#31247)

bcoca committed:
restore connection options from cli

fixes #30720

caphrim007 committed:
Corrects Copyrights and License headers (#31324)

The format has changed for these in the dev_guide. This corrects
them for the F5 modules

s-hertel committed:
Allow any_errors_fatal to be set in playbook - fixes #30691 (#30839)

* Allow any_errors_fatal to be set in playbook.
* Default to the config file value for any_errors_fatal only if it isn't already provided.
* add _get_attr method

mattclay committed:
Avoid msrest version 0.4.15 for Azure modules.

This version has known regressions preventing modules such as
azure_rm_dnsrecordset from working.

bcoca committed:
removed example for restricted facts namespace

(cherry picked from commit e1fab373163ea3213fb51d3460ca5baa7ff29381)

fixes #31330

trishnaguha committed:
nxos_config intended_config docfix (#31337)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

rahushen committed:
fixes #31080 (#31081)

ganeshrn committed:
Fix nxos terminal regex to parse xml response (#31283)

Fixes #26130

Modify nxos terminal regex to parse a xml response.

Commit:
Mention Ansible in package descriptions

These values are used by software updaters to show which applications
are being updates to the user. In the case of Ubuntu, it will now show
clearly that it is updating Ansible instead of a radically simple IT
automation platform.

esuarezsantana committed:
Fix rpm_key absent. (#31045)

When keyid is 8 charactes long, drop_key function cuts everything
issuing an error: "package gpg-pubkey- is not installed"

mkrizek committed:
yum: add integration test for incompatible arch rpm (#31185)

samdoran committed:
Make ansible_selinux facts a consistent type (#31065)

* Make ansible_selinux facts a consistent type

Rather than returning a bool if the Python library is missing, return a dict with one key containing a message explaining there is no way to tell the status of SELinux on the system becasue the Python library is not present.

* Fix unit test

mkrizek committed:
yum: add integration tests for installing from URL (#30616)

bcoca committed:
add note on how to avoid service start in apt

bcoca committed:
use configured ansible_shell_executable

refine args/shell/executable hanlding

gundalow committed:
Improve and link Network Proxy docs (#31316)

* Improve and link Network Proxy docs

It is possible for proxy issues to cause "unable to open shell", so
detail there.

Having the "how-to" in the debug page makes more sense than on the
porting page, so move it.

Fixed a few formatting issues at the same time.

* Edits

* typo

gundalow committed:
data should be redacted in docker_secret (#31366)

trstringer committed:
refactor azure dynamic inventory script configparser import to support python3 (#31370)

* refactor configparser import to support python3

* change ModuleNotFoundError to ImportError

bndabbs committed:
[cloud] EC2_ASG Set desired capacity to min_size if no instances exist (#30987)

* Set desired capacity to min_size if no instances exist

* Improve readability of if/then clause

* Only update null desired_capacity to min_size on initial create

Any future updates to the ASG will be able to reference the existing
capacity.

s-hertel committed:
[cloud] ec2.py: warn that removing a public ipv4 address is not permitted - fixes #30679 (#30952)

* Add a warning to ec2.py if assign_public_ip is different from the matching instance(s) since it isn't a modifiable attribute.

* remove comments

nitzmahone committed:
remove explicit provider reg from azure_rm (#31369)

* now that it's handled automatically as of msrest > 0.4.9

D3DeFi committed:
Allow zabbix_host to change proxy to no proxy, fixes #24735 (#29123)

mvermaes committed:
Fix typo in vault decrypt error message (#31335)

kustodian committed:
Do not run systemctl daemon-reload in check mode

JeanFred committed:
Allow an Elastic IP to be re-associated in ec2_eip (#27389)

Trying to associate an already-associated ElasticIP was failing.

This is however supported by the `boto` method that is used
under the hood, `associate_address`:

To quote `boto` documentation:
```
This option to allow an Elastic IP address that is already
associated with another networkinterface or instance to be
re-associated with the specified instance or interface.
```

This defaults to False, both per backwards-compatibility
and to mirror the boto default value.

Fixes #27385

sduthil committed:
fix misformatted docsite link (#31384)

reason: The link is showing verbatim in the docs, where it should only
show "knowledge base article".

Also, generating the docs shows a lot of:

docs/docsite/rst/win_acl_module.rst:424: WARNING: Unknown target name: "know ledge base article<https://access.redhat.com/articles/rhel-top-support-policies>".

bcoca committed:
fix 'return false' from parse

this was abandoned early on the manger side  but seems like we left behind on plugin side.
more flexible extensions with yaml plugin
validate data correctly for yaml/constructed
fixed issue with only adding one child to keyed, the group only got the host that forced it's creation

fixes #31382
fixes #31365

Etherdaemon committed:
Addition of TCP protocol to ELB target group (#31228)

* Addition of TCP protocol to ELB target group as target groups support HTTP/S and TCP now

* Fixup stickiness type so that it checks if the current_tg has the stickiness_type key in the dict, as TCP ones do not

prasadkatti committed:
Add example showing how to add host to multiple groups (#31152)

* Add example of how to add host to multiple groups

* use 'name' task keyword for comments

jborean93 committed:
win_copy: allow src to have invalid windows chars (#31392)

* win_copy: allow src to have invalid windows chars

* create file with colon instead of avoid checkout errors

eikef committed:
Fix idempotency if zabbix_host when visible_hostname is not set (#31388)

LindsayHill committed:
Fixed RST hyperlink issues (#31394)

rahushen committed:
Fixes #30846 (#30862)

rahushen committed:
nxos_snmp_* fixes : Fixes #30997, #30999 (#31000)

* add codeblock to handle 5K platform for nxos_snmp_host

* fix get_snmp_traps for 5K platform

* Make shippable happy

bcoca committed:
revert ocalhost getting 'all' vars (#31425)

back to previous implicit localhost behaviour

fixes #31420

Nani-o committed:
Adding doc for the description argument of lxd_profile module

abadger committed:
Port arg specs from type='str' to type='path'

abadger committed:
Add a hack to test whether type=path should be used in a module

abadger committed:
Change ping docs to refer to win_ping (#31444)

dagwieers committed:
modprobe: PEP8 compliancy and doc fixes (#31255)

This PR includes:
- PEP8 compliancy fixes
- Documentation fixes
- Make module a bit more consistent with existing practices

abadger committed:
Fix get_exception and wildcard imports for nxos modules

richlv committed:
Update playbooks_lookups.rst

fix doublequote being used in place of the colon

abadger committed:
Remove wildcard imports from cloudstack modules.

Also add boilerplate.

abadger committed:
Update test cases
* Remove cloudstack, f5, and nxos from the wildcard import whitelist
* Remove nxos and f5 from the get_exception whitelist
* Remove some files from the pep8 legacy whitelist

pilou- committed:
ec2_vpc_igw: fix broken import (#31458)

andreaso committed:
Add optional port parameter to nsupdate module (#31446)

An example of when it's useful to be able to specify a custom port
number is when one is using an (ssh) tunnel to update a hidden master.

mkrizek committed:
yum: check if env groups are installed (#31182)

pilou- committed:
ec2_vpc_route_table: fix broken import (#31459)

phy1729 committed:
maven_artifact: Add classifier to default dest (#24518)

abadger committed:
Remove wildcard imports from amazon modules (#31451)

* Remove wildcard imports from amazon modules
* Remove files from legacy pep8 that now pass

ryansb committed:
Properly handle user selection of `None` as vars_files (#31313)

* Properly handle user selection of `None` as vars_files

In a playbook, if a user has a playbook like:

```
- hosts: localhost
  connection: local
  vars_files:
  tasks:
  - ....
```

Then `vars_files` will be none, and cause a `TypeError` in vars-manager when it
tries to iterate over them. To avoid this, I changed the getter to either send
back the vars files from the user, or an empty list when the user passed
`None`.

* Only replace None with an empty list, not all falsey values

* Catch error when vars_files isn't iterable

* Move whole `for` loop into try/except and catch TypeError

* Line length

bcoca committed:
version is string and should be 2.9

jimi-c committed:
Template role_name when statically including via include_role (#31463)

Also trims out code which was incorrectly setting is_static=True for
include_role.

Fixes #30700

abadger committed:
Add boilerplate and remove wildcard imports for openstack modules (#31461)

* Add boilerplate and remove wildcard imports for openstack modules
* Remove openstack files that are now pep8 compliant from the legacy list

pilou- committed:
pip module: fix TypeError (#31395)

* pip: add test: an error occurs when pip not found
* pip: fix TypeError exception when pip executable isn't found

abadger committed:
Remove wildcard import in test_nclu

vtsingaras committed:
win_copy: Add missing check_mode fix (#31469)

* Add missing check_mode fix

* Refactor conditional to use the one embedded in Remove-Item

bcoca committed:
revert namespaced facts changes to package/service

fixes #31485

bcoca committed:
better message for unsupported include/imports

adhoc cannot currently execute these actions

bcoca committed:
warn on bad keys in group

rmfitzpatrick committed:
[cloud] Ensure absence of public_dns_name doesn't halt ec2 module (#31476)

resmo committed:
cs_snapshot_policy: remove CloudStackException dependency

wimnat committed:
[cloud] Fix #31271 - cast all tag values as strings (#31272)

* Fix #31271 - cast all tag values as strings

* Replace `str` with `to_native`

bcoca committed:
validate that existing dest is valid directory

jborean93 committed:
win_copy: added decrypt option (#31291)

jborean93 committed:
added optional environment passthrough for MacOS 10.13 compatibility (#31500)

ozbillwang committed:
New module: Add module for Amazon Systems Manager Parameter Store (cloud/amazon/ssm_parameter_store) (#23460)

- new module: ssm_parameter_store
- new lookup: ssm

* lookup module ssm - adjust error message

* Pacify pylint erroring on botocore not found

* adjust to version 2.5

prasadkatti committed:
Document return values for aws_s3 (#31158)

* Document return values for aws_s3

jborean93 committed:
win_become: Added support to become a service user (#30407)

* win_become: Added support to become a service user

* fixes for linting

* changes to get local and network service working

* fixed linting issues again

* pleasing pepe

jborean93 committed:
updated changelog for win_become with service accounts

arcenik committed:
Fix urlparse import for Python3 (#31240)

* Fix urlparse import for Python3 in

* contrib/inventory/consul_io.py
* contrib/inventory/rudder.py
* contrib/inventory/windows_azure.py
* lib/ansible/module_utils/known_hosts.py
* lib/ansible/modules/cloud/centurylink/clc_firewall_policy.py
* lib/ansible/modules/cloud/docker/_docker.py
* lib/ansible/modules/cloud/ovirt/ovirt_disk.py
* lib/ansible/plugins/action/ce_template.py

bcoca committed:
clarified inventory_dir changes for 2.4 (#31484)

* clarified inventory_dir changes for 2.4

added example on how to restore (mostly) old behaviour for add_hosts

fixes #30901

ar7z1 committed:
minor updates + tests to win_firewall_rule as per jborean93 review (#29148)

* Added warning for 'force' option

* Changed 'profiles' type to list

* Changed 'interfacetypes' type to list

* Added deprecation warning and fixed doc

* updated force parameter

MikeKlebolt committed:
first path check missing forward slash between dcpath and datacenter (#31481)

* fixed folder path logic

* fixed folder path logic

* fixed folder path logic

lukas-bednar committed:
[rpm_key] Fix to import first key on the system (#31514)

* [rpm_key] Fix to import first key on the system

Fixes: #31483

* [rpm_key] removed unsafe_shell and "throwaway" underscore

* [rpm_key] adding test to add the first key on system

mkrizek committed:
yum: fix crashes installing from file/url

arnonki committed:
Use default values for object returned from aws to prevent TypeError  (#31455)

* Use default values for object returned from aws to prevent TypeError
NoneType exception fixes #31454

samdoran committed:
Remove sysctl entries when state=absent (#31486)

* Remove sysctl entry when state=absent
* Cleanup sysctl integration test syntax
* Correct grammar on error message
* Add sysctl integration test for state=absent

s-hertel committed:
aws_s3: fix upload_file's ExtraArgs - fixes #31232 (#31487)

* Module option metadata are extra arguments rather than S3 object metadata: update ExtraArgs variable.

* Remove hyphens from ExtraArgs to maintain backwards compatibility

* Map lowercase extra args to CamelCase

* Maintain backwards compatibility by guessing at content type rather than always defaulting to binary/octet-stream.

* Fix ExtraArgs for non-hyphenated options

* Simplify logic

s-hertel committed:
[cloud] Fix ec2_group when security group lacks a VPC (#31526)

rcarrillocruz committed:
Run neighbors intent tests for ios_interface only if show lldp contains netdev (#31539)

abadger committed:
Try to cleanup after the rpm_key test

nitzmahone committed:
ignore chouseknecht as azure_rm module maintainer

nitzmahone committed:
improve become/runas error messaging on bogus/missing username (#31551)

bcoca committed:
fix paths to actually be config file relative (#31533)

* fix paths to actually be config file relative

also allow to unfrack paths for 'not cwd'
only use -i 'localhost,' for adhoc in pull

bcoca committed:
more updates to plugin/config generation (#30787)

* fixed module generation

added missing lookup page
point to plugins when plugins
made modules singular
add display for verbose an debug messages
nicer templating, changed generation order for ref
corrected links
moved most of lookup docs to plugin section

* Copy edits
* Fixed typos
* Clarified wording

ganeshrn committed:
Fix rollback in junos_config (#31424)

* Fix rollback in junos_config

Fixes #30778

*  Call `load_configuration` with rollback id in case
   the id is given as input
*  Pass rollback id to `get_diff()` to fetch diff from device

* Fix unit test

rcarrillocruz committed:
Introspect the management IP on ios_ping (#31571)

On our CI we don't have external connectivity, so let's ping to
the management interface IP.
Also, ignore errors on the expected failures tests.

rcarrillocruz committed:
Remove duped authorize on ios_ping anchored tasks (#31572)

Otherwise, we get warnings.

rcarrillocruz committed:
Parameterize the ssh port on ios_user tests (#31573)

On CI we use 8022 for SSH, thus we need to explicitly pass
ansible_ssh_port to the ssh command to work.

rahushen committed:
Fix nxos_banner removal idempotence issue in N1 images (#31259)

* Fix nxos_banner removal idempotence issue in N1 images

* handle pep8 error

Commit:
New module: Add monitoring/zabbix_template module. (#25566)

* Implements zabbix_template module.

* Fixes formatting

* Fixes formatting2

* Add templates_clear

* fix some bug

* Fixes formatting

* update ansible 2.5

rahushen committed:
change ports to non well known ports and drop time_range for N1 (#31261)

pdellaert committed:
Fix #21112 : Adding support for configuring a CD-rom iso image (#28155)

* Initial CD-ROM support

* create cdrom bugfix

* Improving CDROM change detection and fixing template creation bug
Running MarkAsTemplate on an existing template will fail with an error

* Better change detection for guest ID
Should only mark a change in case it actually changes

* Adding integration tests

* Pep8 compliance fixes

* Adding CDROM support, including iso, client and none types

* Updating added release version for CDROM option

voipmeister committed:
Changed src_intf and dst_intf to list type (#31479)

ganeshrn committed:
Remove log of command executed in persistent connection (#31581)

resmo committed:
cs_snapshot_policy: fix VM not found in non-default zone

resmo committed:
tests: cs_snapshot_policy: add integration tests

resmo committed:
cs_securitygroup: remove CloudStackException dep (#31597)

- fixes pep8
- fixes docs

candlerb committed:
Add ellipsis to indicate YAML example is excerpt, not standalone

jimi-c committed:
Fix static determination for include_tasks

An incorrect removal of a conditional resulted in include_tasks falling
through to the old static detection mechanism incorrectly. This restores
the previous conditional check.

Fixes #31593

resmo committed:
cs_securitygroup_rule: remove CloudStackException dependency (#31603)

- fixes pep8
- fixes docs

schast committed:
zabbix_host: added zabbix host property (description) (#25969)

* zabbix_host: added zabbix host property (description)

* zabbix_host: fixed error E309 version_added for new option (description) should be 2.4

* zabbix_host: deleted unwanted else for update description

* lib/ansible/modules/monitoring/zabbix_host.py: increased version_added to 2.5 for option description

* lib/ansible/modules/monitoring/zabbix_host.py: allow to change the description

* lib/ansible/modules/monitoring/zabbix_host.py: added new lines back to fix pep8 issues

caphrim007 committed:
Adds new committer to Ansible core (#31544)

This is me (Tim) signing and agreeing to the committer guidelines.

jborean93 committed:
windows: removed #Requires from PS pipelines as it causes STDERR false positives (#31558)

pweldon committed:
Fix typo in command output of encrypt_string example. (#31601)

resmo committed:
cs_portforward: remove CloudStackException dep (#31610)

- fixes pep8
- fixes doc

mattclay committed:
Work around cryptography 2.1 requiring pip 8.1.2+

machacekondra committed:
ovirt_templates: Add seal parameter (#31482)

abadger committed:
Remove boundary_meter module as the hosted service it configured has been removed. (#31506)

References #29387

colin-nolan committed:
Corrects examples in the `find` module. (#31522)

cognifloyd committed:
Fix Inventory Plugin Loading (#31605)

This change makes the PluginLoader use DEFAULT_INVENTORY_PLUGIN_PATH setting.
Inventory Plugins were only being loaded the 'inventory_plugins' folder of the current directory,
as well as the ansible-provided inventory plugins (e.g. `/path/to/site-packages/ansible/plugins/inventory`).

prasadkatti committed:
[aws_s3] different fail msg depending on whether version is specified (#31598)

mattclay committed:
Update Azure tests to limit cryptography version. (#31616)

* Revert "Work around cryptography 2.1 requiring pip 8.1.2+"

This reverts commit a10a80b8a36192982829aefbba553dfdc6d9aa55.

* Update Azure tests to limit cryptography version.

gundalow committed:
botmeta for lib/ansible/plugins/filter/network.py (#31608)

resmo committed:
cs_instance_facts: fix pep8 (#31606)

resmo committed:
botmeta: maintain cloudstack module utils (#31613)

trishnaguha committed:
fix ios_interface test unsupported param (#31622)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

Akasurde committed:
pep8 fixes for vmware modules (#31537)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

trishnaguha committed:
fix nxos_vlan unexpected deprecation warning (#31624)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

mattclay committed:
Remove cryptography install from CI other test.

mattclay committed:
Multiple Parallels host support for ansible-test. (#31552)

Commit:
Create action file enos.py (#31508)

* Create action file enos_facts.py

* Update and rename enos_facts.py to enos.py

* Taking chances on Dealing with Unstable issues

* Removing blank space/ white line

Commit:
create cliconf plugin for enos - enos.py (#31509)

Commit:
create terminal plugin for enos modules (#31510)

Commit:
Create document fragment for enos modules (#31512)

Commit:
Update base.yml to add enos from lenovo (#31565)

* Update base.yml to add enos from lenovo

* Update base.yml

Commit:
Adding enos/__init__.py (#31580)

##### SUMMARY
Creating the modules for enos from Lenovo.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/network/enos/__init__.py

##### ANSIBLE VERSION
ansible 2.5.0
config file = /etc/ansible/ansible.cfg
configured module search path = [u'/root/.ansible/plugins/modules',
u'/usr/share/ansible/plugins/modules']
ansible python module location =
/usr/local/lib/python2.7/dist-packages/ansible-2.5.0-py2.7.egg/ansible
executable location = /usr/local/bin/ansible
python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

##### ADDITIONAL INFORMATION

This is an effort to create modules for enos based switches from Lenovo.

machacekondra committed:
ovirt_clusters: add alias performance_preset for memory_policy (#31523)

machacekondra committed:
ovirt_networks: Add label to network (#31517)

machacekondra committed:
ovirt_templates: Fix exlusive parameter (#31470)

machacekondra committed:
ovirt_vms: Wait for VM disks when adding VM (#31468)

Fixes: https://github.com/ansible/ansible/issues/30515

bcoca committed:
highlight dynamic parts of error msg

jctanner committed:
Only install pyvmomi if user is root (#31628)

ZubairLK committed:
vmware_vm_facts: Add error check for config.vm (#31629)

Add a simple check to see if vm.config exists before looking for
vm.config.hardware

MihaiMoldovanu committed:
etcd.py module improved . added posibility to configure from playbook… (#18580)

* etcd.py module improved . added posibility to configure from playbook etcd server to query

* Update playbooks_lookups.rst

documented etcd lookups

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Fixed errors reported by ansibot

* Update playbooks_lookups.rst

* Update playbooks_lookups.rst

* Copy edits.

sethp-nr committed:
wait_for: treat broken connections as "unready" (#28839)

* wait_for: treat broken connections as "unready"

We have observed the following condition while waiting for hosts:

```
Traceback (most recent call last):
  File "/var/folders/f8/23xp00654plcv2b2tcc028680000gn/T/ansible_8hxm4_/ansible_module_wait_for.py", line 585, in <module>
    main()
  File "/var/folders/f8/23xp00654plcv2b2tcc028680000gn/T/ansible_8hxm4_/ansible_module_wait_for.py", line 535, in main
    s.shutdown(socket.SHUT_RDWR)
  File "/usr/local/opt/python/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 57] Socket is not connected
```

This appears to happen while the host is still starting; we believe something is
accepting our connection but immediately resetting it. In these cases, we'd
prefer to continue waiting instead of immediately failing the play.

This patch has been applied locally for some time, and we have seen no adverse
effects.

* wait_for: fixup change

We were missing an import and a space after the `#`

kairoaraujo committed:
new module: AIX rootvg backup image using mksysb (#30460)

* new module: AIX rootvg backup image using mksysb

This module is simple but very useful for AIX system
administrators. Easy to construct playbooks to generate
and manage rootvg backups using mksysb tool.

* added module_check, pep8, non-written convention

- implemented module_check;
- fixed some pep8 and non-written convention

* removed parameters as global variables and doc

Moved global variables parameters to inside main()
Better doc format for mentioned files

machacekondra committed:
ovirt: Add new api facts module (#28875)

* ovirt: Add new api facts module

* Simplify boilerplate and add Python 3 import

pilou- committed:
s3_sync: fix broken import (#31460)

bcoca committed:
update delegated var fetching for new 'loop' (#31588)

* update delegated var fetching for new 'loop'

* mixed loop/terms

tedder committed:
[cloud][docs] Document default/required 'state' in CloudFormation module (#31642)

* fix documentation to match default/required status of 'state'

* add 'choices'

sampritavh committed:
[cloud] Support changeset_name parameter on CloudFormation stack create (#31436)

mattclay committed:
Remove cryptography version limit in Azure tests.

No longer needed now that cryptography 2.1.1 has been released.

jmighion committed:
Adding a cli transport option for the bigip_command module. (#30391)

* Adding a cli transport option for the bigip_command module.

* Fixing keyerror when using other f5 modules. Adding version_added for new option in bigip_command.

* Removing local connection check because the F5 tasks can be delegated to any host that has the libraries for REST.

* Using the network_common load_provider.

* Adding unit test to cover cli transport and updating previous unit test to ensure cli was not called.

caphrim007 committed:
Adds a refactored bigip_monitor_http module. (#30998)

This patch refactors the bigip_monitor_http module to use REST
instead of SOAP. It additionally adds unit tests and current F5
code conventions.

Integration tests can be found here
* https://github.com/F5Networks/f5-ansible/blob/devel/test/integration/bigip_monitor_http.yaml
* https://github.com/F5Networks/f5-ansible/tree/devel/test/integration/targets/bigip_monitor_http/tasks

jborean93 committed:
update Windows docs (#30783)

* first round of splitting and expanding Windows docs

* Added info about WinRM and started the usage page

* more details onf usage around command/shell

* added more info

* missed a page

* added info about hotfix and ps upgrade script

* Added DSC information

* filled in FAQ and fixed up links

* changes per PR review and typo fixes found by myself

* fixed sanity check

* updated external script URLS

* Edits

* Edits

* Edits

* Copy edits.

* More copy edits.

* Added info about becoming a local service account

* some formatting changes on become and DSC indent fix

* Copy edits/rewrites

* Copy edits/rewrites

* minor updates to UAC text, typos

* minor wording updates

* minor wording tweaks/typos

* minor wording tweaks/typos

* minor tweaks/typos

* fix indent issue

* fix title underline length

* fix quote issue

ryansb committed:
Rename ssm_parameter_store module and ssm lookup to be prefixed with aws (#31662)

wxdao committed:
Fix typo in playbooks_error_handling (#31636)

rahushen committed:
Fixes #31056 (#31057)

Lujeni committed:
pip: Use chdir directive in the venv path (fixes #25122) (#26865)

mensler committed:
Update Zabbix maintenance if it already exists to set new period, hosts etc. (#24371)

* Update Zabbix maintenance if it already exists to set new period, hosts etc.

* Whitespace

kevinburke committed:
Un-nest nested if statements (#19190)

Easier and clearer to write

if a and b:
    foo()

than

if a:
    if b:
        foo()

Lorquas committed:
updating argument validation for activation keys to match subscription-manager command line, fixes #27283 (#30236)

Akasurde committed:
Correct usage for shutil.rmtree (#31541)

* Correct usage for shutil.rmtree

Fix adds correct usage of shutil.rmtree in git module

Fixes: #31225

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

* Include archive tests so they get run

* Use new include syntax

* Cleanup syntax on git tests

- use multi-line YAML
- remove unneeded {{ }} around vars in conditionals
- remove unneeded quotes
- add task file name to task names for easier troubleshooting when things fail

* Make archive tests work for RHEL/CentOS 6

The older versions of Jinja2 in RHEL/CentOS 6 required assertion tasks using the map filter to be skipped.

The older version of git required gzip compression to be skipped on RHEL/CentOS 6.

* Account for ansible_distribution_major_version missing

KKoukiou committed:
cloud: ovirt: Add ovirt_disk_facts module (#30422)

ganeshrn committed:
Validate parse_cli filter inputs (#31293)

* Validate parse_cli filter inputs

Fixes #30517

*  Add check to validate input is of type string
*  Add check to confirm template file exist

* Update error message for parse_cli_textfsm invalid template path

* Add input validation for parse_cli_textfsm filter

caphrim007 committed:
This patch fixes a number of outstanding bugs and code convention problems. (#31618)

* documentation was not inline with other Ansible modules
* Python 3 specific imports were missing
* monitor_type is no longer required when creating a new pool; it is now the default.
* A new monitor_type choice of "single" was added for a more intuitive way to specify "a single monitor". It uses "and_list" underneath, but provides additional checks to ensure that you are specifying only a single monitor.
* host and port arguments have been deprecated for now. Please use bigip_pool_member instead.
* 'partition' field was missing from documentation.
* A note that "python 2.7 or greater is required" has been added for those who were not aware that this applies for ALL F5 modules.
* Unit tests were fixed to support the above module

sivel committed:
Port vault-keyring.py to properly load the config. Fixes #31304 (#31315)

kennethgillen committed:
Fix gender pronouns and the odd typo (#31709)

s-hertel committed:
[cloud] Document alias for aws_s3 'overwrite' option. (#31638)

s-hertel committed:
[cloud] Require src parameter when aws_s3 is used with mode 'put' (#31657)

s-hertel committed:
[cloud] Don't try to deregister an instance that isn't in the load balancer (#31660)

s-hertel committed:
[cloud] Only enforce state (running/stopped/etc) in EC2 "exact_count" when state is specified (#31648)

fix issue in ec2 module where exact count would create new instances if the instane state is stopped or terminated

alikins committed:
Vault secrets script client inc new 'keyring' client (#27669)

This adds a new type of vault-password script  (a 'client') that takes advantage of and enhances the 
multiple vault password support.

If a vault password script basename ends with the name '-client', consider it a vault password script client. 

A vault password script 'client' just means that the script will take a '--vault-id' command line arg.

The previous vault password script (as invoked by --vault-password-file pointing to an executable) takes
no args and returns the password on stdout. But it doesnt know anything about --vault-id or multiple vault
passwords.

The new 'protocol' of the vault password script takes a cli arg ('--vault-id') so that it can lookup that specific
vault-id and return it's password.

Since existing vault password scripts don't know the new 'protocol', a way to distinguish password scripts
that do understand the protocol was needed.  The convention now is to consider password scripts that are
named like 'something-client.py' (and executable) to be vault password client scripts.

The new client scripts get invoked with the '--vault-id' they were requested for. An example:

     ansible-playbook --vault-id my_vault_id@contrib/vault/vault-keyring-client.py some_playbook.yml

That will cause the 'contrib/vault/vault-keyring-client.py' script to be invoked as:

     contrib/vault/vault-keyring-client.py --vault-id my_vault_id

The previous vault-keyring.py password script was extended to become vault-keyring-client.py. It uses
the python 'keyring' module to request secrets from various backends. The plain 'vault-keyring.py' script
would determine which key id and keyring name to use based on values that had to be set in ansible.cfg.
So it was also limited to one keyring name.

The new vault-keyring-client.py will request the secret for the vault id provided via the '--vault-id' option.
The script can be used without config and can be used for multiple keyring ids (and keyrings).

On success, a vault password client script will print the password to stdout and exit with a return code of 0.
If the 'client' script can't find a secret for the --vault-id, the script will exit with return code of 2 and print an error to stderr.

bcoca committed:
additional configmanager tests

left placeholders for more

mattclay committed:
Fix Debian packaging Dockerfile and docs.

resmo committed:
cs_instance: remove CloudStackException dep (#31724)

- fixes dep
- fix docs

fabianvf committed:
use to_str instead of json.dumps when serializing k8s object for logging

fabianvf committed:
Update k8s_common.py

caphrim007 committed:
Fixes various gtm pool issues (#31728)

Various formatting related fixes. Also fixed an idempotency problem
with the 'disabled' state

caphrim007 committed:
Fixes documentation related bugs (#31730)

New conventions for ansible warrant fixes to accomodate those
in bigip_partition.

This patch also includes an import fix that can raise an error when
Ansible unit tests run

caphrim007 committed:
Adds various provision fixes (#31731)

* vcmp provisioning support
* documentation fixes
* fixes for python3 causing an exception

caphrim007 committed:
Refactors bigip_selfip (#31732)

In this refactor we moved to the most recent coding standards for
both F5 and Ansible. Many bugs were fixed and some features were
also added (such as ipv6 support).

caphrim007 committed:
Fixes an fqdn error (#31745)

the pool module was not correctly interpreting the partitions
in the monitors list vs the partition at the task level. This patch
fixes that.

yaacov committed:
fix the handeling of empty auth_key (#31753)

Commit:
Update ssm lookup to support getting parameters by path (#31752)

eikef committed:
Fix #24964 comparaison string vs integer (#31701)

machacekondra committed:
ovirt_networks: Fix label assigment condition (#31772)

trishnaguha committed:
Fixes #30529 nxos_facts ipv6 interface (#31286)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

prasadkatti committed:
[docs] Clean up the cloudformation examples (#31770)

- added `role_arn` to the "role example" example
- removed the irrelevant parameters to the "role example" example
- updated comment on one of the examples
- removed the last example as it was a duplicate of "role example" example
- some other minor changes

fantashley committed:
Make RPM spec compatible with RHEL 6 (#31653)

* Make RPM spec compatible with RHEL 6

* Define __python2 macro only when undefined

Commit:
zabbix: add validate certs param (#31739)

willthames committed:
Add pagination for IAM group API calls (#31554)

get_group and list_attached_policies_for_groups can be paginated
so add pagination handling.

Fixes #31545

bcoca committed:
better cleanup on task results display (#27175)

* better cleanup on task results display

callbacks get 'clean' copy of result objects
moved cleanup into result object itself
removed now redundant callback cleanup
moved no_log tests

* moved import as per feedback

evancarter-iex committed:
gce_instance_template: fix allow the setting of the service account (#29154)

kedarkekan committed:
Update BOTMETA.yml

fungusakafungus committed:
Fix -M/--module-path option for ansible-doc and ansible-console (#31744)

Fixes #31743

gregorydulin committed:
Update --extra-vars documentation. Fixes #11647 (#31433)

* Update --extra-vars documentation to address #11647

* Fix some grammar

* Fix indentation

* Reshuffle and add versionadded tags

* Remove code-block tags

* Add lf after versionadded lines

Commit:
Add sf-facts module.

Commit:
Merge branch 'sf-facts' of github.com:hulquest/ansible into sf-facts

hulquest commented:
Closing this PR since I messed up the rebase.  I'll reopen a new one.
